### PR TITLE
Eco 457/add native spend bi events

### PR DESCRIPTION
**Pull request by YohayBar**
WIP

### CODE REVIEW

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 1]
```diff
@@ -22,16 +26,18 @@
...
 	private final String orderJwt;
 	private final ExternalOrderCallbacks externalOrderCallbacks;
+	private final EventLogger eventLogger;
 
 	private OpenOrder openOrder;
 	private MainThreadExecutor mainThreadExecutor = new MainThreadExecutor();
 
 	CreateExternalOrderCall(@NonNull OrderDataSource.Remote remote, @NonNull BlockchainSource blockchainSource,
-		@NonNull String orderJwt,
+		@NonNull String orderJwt, EventLogger eventLogger,
```
Comment (OrenZak) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: should be annotated as `@NonNull`?

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 2]
```diff
@@ -52,18 +60,14 @@ public void run() {
...
-					}
-				});
+				onOrderFailed(ErrorUtil.fromApiException(e));
 			}
 			return;
 		}
 
 		if (externalOrderCallbacks instanceof ExternalSpendOrderCallbacks) {
 			// Send transaction to the network.
+			sendSpendOrderCompletionSubmittedEvent();
```
Comment (OrenZak) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: can be removed, it's on purchase callbacks

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 3]
```diff
@@ -100,6 +104,33 @@ public void run() {
...
+					eventLogger.send(SpendOrderCreationReceived
+						.create(openOrder.getOfferId(), openOrder.getId(), true));
+					break;
+				case EARN:
+					break;
+			}
+		}
+	}
+
+	private void sendSpendOrderCompletionSubmittedEvent() {
```
Comment (OrenZak) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: can be removed

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 4]
```diff
@@ -129,35 +160,44 @@ public void onResponse(final Order order) {
...
-				});
+				onOrderFailed(ErrorUtil.fromApiException(e));
 			}
 		}).start();
 	}
 
-	private void runOnMainThread(Runnable runnable) {
-		mainThreadExecutor.execute(runnable);
-	}
+	private void onOrderFailed(final KinEcosystemException exception) {
```
Comment (OrenZak) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: `onOrderFailed` this method being called if: `createExternalOrderSync` failed and also from `getOrder` failed.
but you are sending only `SpendOrderCreationFailed` should be `SpendOrderFailed` on `getOrder` failed method.

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/OrderRepository.java [hunk 5]
```diff
@@ -299,59 +303,86 @@ public void onFailure(ApiException e) {
...
+				}
 
-				});
+				@Override
+				public void onTransactionFailed(final OpenOrder openOrder, final KinEcosystemException exception) {
+					cancelOrder(openOrder.getOfferId(), openOrder.getId(), new KinCallback<Void>() {
+						@Override
+						public void onResponse(Void response) {
+							handleOnFailure(exception, openOrder != null ? openOrder.getOfferId() : "null",
+								openOrder != null ? openOrder.getId() : "null");
```
Comment (OrenZak) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/OrderRepository.java]: `openOrder` params can't be null at this point.

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/OrderRepository.java [hunk 6]
```diff
@@ -299,59 +303,86 @@ public void onFailure(ApiException e) {
...
+						}
 
-			}
+						@Override
+						public void onFailure(KinEcosystemException e) {
+							handleOnFailure(exception, openOrder != null ? openOrder.getOfferId() : "null",
+								openOrder != null ? openOrder.getId() : "null");
```
Comment (OrenZak) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/OrderRepository.java]: same here

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 7]
```diff
@@ -100,6 +101,36 @@ public void run() {
 		});
 	}
 
+	private void sendOrderCreationFailedEvent(ApiException exception) {
+		if (openOrder != null) {
```
Comment (YohayBar) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: need to send even if openOrder !=null

Comment (YohayBar) [reply kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: we have a problem - if the SpendOrderCreationFailed it means that we don't have openOrder == null by definition ????

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 8]
```diff
@@ -129,35 +160,32 @@ public void onResponse(final Order order) {
...
+		runOnMainThread(new Runnable() {
+			@Override
+			public void run() {
+				externalOrderCallbacks
+					.onOrderFailed(exception, finalOpenOrder);
+			}
+		});
 	}
 
-	private String getApiExceptionsMessage(Throwable t) {
```
Comment (YohayBar) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: wasn't used?

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 9]
```diff
@@ -51,19 +58,13 @@ public void run() {
...
 			} else {
-
-				runOnMainThread(new Runnable() {
-					@Override
-					public void run() {
-						externalOrderCallbacks.onOrderFailed(ErrorUtil.fromApiException(e));
-					}
-				});
+				sendOrderCreationFailedEvent(e);
+				onOrderFailed(ErrorUtil.fromApiException(e));
```
Comment (YohayBar) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: we will not have information regarding the order - if we arrived to the catch it means the openOrder is null

File: kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java [hunk 10]
```diff
@@ -129,35 +160,32 @@ public void onResponse(final Order order) {
...
+		});
 	}
 
-	private String getApiExceptionsMessage(Throwable t) {
-		try {
-			return ((ApiException) t).getResponseBody().getMessage();
-		} catch (Exception e) {
-			return hasMessage(t) ? t.getMessage() : "Task failed";
-		}
+	private void runOnMainThread(Runnable runnable) {
```
Comment (YohayBar) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: @OrenZak ??

Comment (YohayBar) [inline kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/data/order/CreateExternalOrderCall.java]: ??? @OrenZak

**Pull request merged**